[#19] 로그인 체크 기능 구현

msugo1 opened the pull request:
### HandlerInterceptor & `@LoginRequired`

- 인터셉터와 어노테이션을 활용해 현재 사용자가 로그인 했는지 먼저  확인하는 기능을 구현

- ArgumentResolver와 `@CurrentUser`를 세션에 있는 아이디를 주입하기 전  먼저 로그인 여부를 확인

### UnauthorizedAccessException

- 로그인하지 않은 사용자가 로그인이 필요한 기능에 접근 시 예외를  던지는데 필요한 클래스

### ControllerExceptionHandler

- 위의 예외를 로그인 관련 ExceptionHandler 메소드에 매핑 & 문구 수정

- `friendRequestException` 메소드가 반환하는 예외 `title`이 좀 더  범용적으로 활용될 수 있도록 변경

### WebConfig

- 위에서 작성한 `HandlerInterceptor` 추가

### FriendController, LanguageController, MyPageController

- 각 컨트롤러 메소드가 `@CurrentUser`을 파라미터로 주입 받고 있는 경우  `@LoginRequired`를 붙여서 인터셉터가 동작하도록 설정

- FriendController에서 불필요하게 들어간 `@ResponseStatus(HttpStatus.OK)` 모두 삭제

Commit:
[#19] 로그인 체크 기능 구현

* HandlerInterceptor & `@LoginRequired`

- 인터셉터와 어노테이션을 활용해 현재 사용자가 로그인 했는지 먼저
  확인하는 기능을 구현

- ArgumentResolver와 `@CurrentUser`를 세션에 있는 아이디를 주입하기 전
  먼저 로그인 여부를 확인 하기 위함

* UnauthorizedAccessException

- 로그인하지 않은 사용자가 로그인이 필요한 기능에 접근 시 예외를
  던지는데 필요한 클래스

* ControllerExceptionHandler

- 위의 예외를 로그인 관련 ExceptionHandler 메소드에 매핑 & 문구 수정

- `friendRequestException` 메소드가 반환하는 예외 `title`이 좀 더
  범용적으로 활용될 수 있도록 변경

* WebConfig

- 위에서 작성한 HandlerInterceptor 등록

* FriendController, LanguageController, MyPageController

- 각 컨트롤러 메소드가 `@CurrentUser`을 파라미터로 주입 받고 있는 경우
  `@LoginRequired`를 붙여서 인터셉터가 동작하도록 설정

- FriendController에서 불필요하게 들어간 `@ResponseStatus(HttpStatus.OK)` 모두 삭제

msugo1 commented:
1. 예외 관련 커스텀 클래스가 지나치게 많아지는 것 같은데 크게 상관 없는 부분일까요? 좀 줄여볼 필요가 있다면 어떤 방법을 활용할 수 있을까요?

2. https://soolow-and-sooteady.tistory.com/4

`Filter vs Interceptor vs AOP`를 나름대로 비교해서 선택한 블로그 글도 작성완료 했습니다. 

시간나실 때 저번에 `Scale up vs Scale out` 그거랑 해서 피드백 해주시면 감사하겠습니다!

f-lab-dev commented on src/main/java/me/soo/helloworld/util/handler/ControllerExceptionHandler.java:
```diff
@@ -31,10 +31,13 @@
+    public ResponseEntity<ExceptionResponse> loginFailedException(final RuntimeException ex) {
         log.error(ex.getMessage(), ex);
-        ExceptionResponse response = new ExceptionResponse("로그인에 실패하였습니다.", ex.getMessage());
+        ExceptionResponse response = new ExceptionResponse("로그인 관련 오류입니다.", ex.getMessage());
```
`로그인에 실패한것`과 `인증되지 않은 상태로 인증이 필요한 API에 접근하는것`은 따로 분리하는게 좋지 않을까요?

msugo1 replied:
`ExceptionHandler` 내에 메소드가 좀 많아진 감이 있어서 줄여보려고 했던 건데 생각해보니 성격이 다르네요. 분리해서 적용했습니다.

f-lab-dev commented on src/main/java/me/soo/helloworld/util/CheckLoginInterceptor.java:
```diff
@@ -0,0 +1,28 @@
+    @Override
+    public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler) throws Exception {
+        HandlerMethod handlerMethod = (HandlerMethod) handler;
+        String userId = (String) request.getSession().getAttribute(USER_ID);
```
아래의 부분은 캡슐화된 `LoginService`를 사용하는게 낫지 않을까요?
`(String) request.getSession().getAttribute(USER_ID);`

msugo1 replied:
거기까지는 미처 생각을 못했네요. `LoginService`를 주입받아서 해당 클래스에 있는 `getCurrentUserId()`를 사용하도록 변경했습니다.

Commit:
[#19] 피드백 반영

1. LoginInterceptor의 preHandle

- 메소드 내에서 현재 로그인 중인 사용자의 아이디를 세션으로 받을 때
  캡슐화가 되어 있는 `LoginService`의 `getCurrentUserId()`메소드를
  사용하는 것으로 변경

2. ExceptionHandler내의 로그인 실패/인증 예외 분리

- 합쳐져 있던 로그인 실패와 인증 관련 예외를 분리

f-lab-dev reviewed (approved):
예외가 `me.soo.helloworld.exception` 패키지에 몰려있어서 많아보일 수 있지만 패키지를 더 세밀히 나눠주면 구분이 좀 잘 갈것같습니다~

블로그는 보고 피드백 드리겠습니다~
